src/networks/infinitesquarenetwork.jl

```diff
@@ -74,42 +91,50 @@ function Base.isapprox(A₁::InfiniteSquareNetwork, A₂::InfiniteSquareNetwork;
 end
 
 ## Rotations
-function Base.rotl90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}} # Rotations of matrix unit cells
-    return NWType(rotl90(rotl90.(unitcell(A))))
-end
-function Base.rotr90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}}
-    return NWType(rotr90(rotr90.(unitcell(A))))
-end
-function Base.rot180(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,2}}
-    return NWType(rot180(rot180.(unitcell(A))))
-end
-function Base.rotl90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}} # Rotations of cubic unit cells along z-axis
-    return NWType(stack(rotl90, eachslice(unitcell(A); dims=3)))
+
+function Base.rotl90(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rotl90(_rotl90_localsandwich.(unitcell(n))))
 end
-function Base.rotr90(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}}
-    return NWType(stack(rotr90, eachslice(unitcell(A); dims=3)))
+function Base.rotr90(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rotr90(_rotr90_localsandwich.(unitcell(n))))
 end
-function Base.rot180(A::NWType) where {NWType<:InfiniteSquareNetwork{<:Any,3}}
-    return NWType(stack(rot180, eachslice(unitcell(A); dims=3)))
+function Base.rot180(n::InfiniteSquareNetwork)
+    return InfiniteSquareNetwork(rot180(_rot180_localsandwich.(unitcell(n))))
 end
 
-## OptimKit optimization compatibility
-function LinearAlgebra.rmul!(A::InfiniteSquareNetwork, α::Number) # Used in _scale during OptimKit.optimize
-    rmul!.(unitcell(A), α)
-    return A
+## Chainrules
+
+function ChainRulesCore.rrule(
+    ::typeof(Base.getindex), network::InfiniteSquareNetwork, args...
+)
+    O = network[args...]
+
+    function getindex_pullback(ΔO_)
+        ΔO = map(unthunk, ΔO_)
```

Comment:
this looks a bit strange to me, what if `args... === i::Int`? or what if `args... === Colon()`? Do these methods work the same for both? I would somehow expect that this method only needs to be defined for the scalar getindex?

This being said, since we implement `getindex` through `getindex(unitcell())`, a `rrule` for `unitcell` should suffice, and then this could also be deleted?

Reply:
Yes, we only really need the scalar getindex, I'll specialize this to avoid issues.

We could just go through a `unitcell` `rrule`, but out of the box this would get in a matrix of weird tangents that I would have to convert before being able to construct an `InfiniteSquareNetwork` from this. I can give this a shot, but at first sight I don't know if that's going to be actually cleaner than doing it this way.